### ISSUE TITLE
Symbols not loaded for current project according to tsconfig.json lookup logic (only for first tsconfig.json found in project)

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/TypeScript#8571]
_From @ashelley on May 11, 2016 23:4_

- VSCode Version: 1.10
- OS Version: Windows 10

Symptom: 

VSCode only loads symbols for the first tsconfig.json it finds in the directory tree.  If you try to look up a symbol using Ctrl+T you will only see symbols from the first tsconfig.json in the project.

Expected results: 

When hitting Ctrl+T it should look at the current active file and find the next tsconfig.json up the directory tree and use that to load symbols.  Currently it seems that the syntax checker (compiler) uses the right logic to determine the correct symbols to load but Ctrl+T does not.

Steps to Reproduce:

1. clone this sample repo https://github.com/ashelley/vscode-tsconfig-all-symbols-test.git
2. Navigate to any project that is not the first project in the list (module3/Biz.ts)
3. Notice that symbols are loaded for only the first project:

![image](https://cloud.githubusercontent.com/assets/3195045/15199261/8b188b04-1791-11e6-9bf6-471319d8402c.png)

4. Try to load a symbol from the current project.  Notice how the symbol loads for the syntax checker (compiler) (no errors) but does not load the symbol when hitting ctrl+T

![image](https://cloud.githubusercontent.com/assets/3195045/15199286/c17d4eb4-1791-11e6-86c5-72c6afea9eb7.png)

![image](https://cloud.githubusercontent.com/assets/3195045/15199292/d5a4a9be-1791-11e6-9e70-79363600d24f.png)

_Copied from original issue: Microsoft/vscode#6300_

**Comment by dbaeumer** [comment]
_From @ashelley on May 11, 2016 23:11_

Another example of a discrepancy between the symbol lookup and other features of vscode is that peek definition still works (even if the symbol lookup does not)

 
![image](https://cloud.githubusercontent.com/assets/3195045/15199440/f62bf164-1792-11e6-8f53-a40d209837f3.png)

**Comment by dbaeumer** [comment]
Due to a limitation of the tsserver workspace symbols are currently scoped to the project the focused text editor is part of.

I will move this issue to TS since it requires work on the tsserver side.

**Comment by ashelley** [comment]
@dbaeumer just to clarify:

"Due to a limitation of the tsserver workspace symbols are currently scoped to the project the focused text editor is part of."

This is the expected result.  

However the current behavior loads symbols not for the currently focused project but of the first project in the current workspace.

**Comment by mhegazy** [comment]
This works as i would expect. a file can have only one "automatically detected" tsconfig.json. there could be other tsconfig.json files that include the file, and there is no easy way to know them all, nor it is clear why would you want to search them all. i would suggest having one tsconfig.json at the root for your project.

**Comment by ashelley** [comment]
@mhegazy I guess it's just confusing because peek definition and the type error highlighting work properly.  In other words these features don't rely on the "one automatically detected" tsconfig.json.  

However.... Ideally the symbol lookup would load all tsconfigs detected in the project similar to how visual studio works when you have multiple projects in a solution file.

**Comment by mhegazy** [comment]
They are not the same symbol. so not sure what you mean by "how visual studio works". the project `module3` only has 3 files, and there is no "context" where `MyProhect` namespace has `Foo`, `Bar` and `Biz`. so not sure i understand what is the request here.

**Comment by ashelley** [comment]
Okay, possibly my fault for not explaining properly... Ignore the visual studio comment for now. 

I'm hoping this graphic can explain why I think this is a bug an unexpected behavior as originally reported:

![image](https://cloud.githubusercontent.com/assets/3195045/15368076/e0014102-1ce0-11e6-9293-8f00bedb475d.png)

Notice how the symbol drop down does not load symbols for the current project but yet the current project understands that Bar exists.  If the project didn't think Bar existed there would be a syntax error.  As shown here (using non existent Baz) variable:

![image](https://cloud.githubusercontent.com/assets/3195045/15368210/7ec99f46-1ce1-11e6-9250-ca54e64f6330.png)

To make the problem worse.. by taking a naive approach of loading symbols  from "the first tsconfig.json"  and not following the same semantics as what the actual compiler does you run into an even more confusing situation where the symbol lookup shows symbols exist even if they are not available to the current project.  (Again the compiler and type checking work properly)

To demonstrate this I create a non exported type in the "first project":

![image](https://cloud.githubusercontent.com/assets/3195045/15368376/51417e6c-1ce2-11e6-93d3-0513c489f89a.png)

And try to use it in the current project.  Notice how the compiler/type checker recognize this as wrong but the symbol table still shows it.

![image](https://cloud.githubusercontent.com/assets/3195045/15368433/953e7b38-1ce2-11e6-8c88-dbb2420b032e.png)

In other words... my request is that the symbols from "ctrl+t" lookup dropdown match the behavior of the rest of vscode (type checking and peek definition).

Alternatively... if symbols could be indexed from all projects this would make it even more useful.  

Currently the symbol lookup only works if you only have on tsconfig.json.  This is not terrible but it seems to me that the behavior of symbol lookup should match the behavior of typechecking and peek definition 1) to be more useful, 2) to avoid confusing situations where it shows types and the types don't exist.

**Comment by mhegazy** [comment]
> Currently the symbol lookup only works if you only have on tsconfig.json. This is not terrible but it seems to me that the behavior of symbol lookup should match the behavior of typechecking and peek definition 1) to be more useful, 2) to avoid confusing situations where it shows types and the types don't exist.

I do not see the same behavior in latest:

![animation](https://cloud.githubusercontent.com/assets/8000722/15376934/03ea9e66-1d0d-11e6-8e88-a0bdfca84519.gif)

> confusing situation where the symbol lookup shows symbols exist even if they are not available to the current project

The symbol search is to search for "named" declarations, regardless of their scope; this is "jump to declaration" where ever it is. 

Completion is what shows you available symbols in the current scope. completion will not show you the type. symbols each will.

**Comment by ashelley** [comment]
thanks for looking at this more closely @mhegazy 

I believe you are currently seeing what "appears to be" the correct behavior only because you happen to open vscode with Biz.ts as the last remembered open file.  So in this situation for that particular module you will see the correct symbols loaded because you were lucky enough to get the right tsconfig.json file loaded by chance because you just happened to be in the right project when you closed/reopened vscode.

If you simply select another file in module 1 or module 2 close and reopen visual studio code and attempt the same behavior you will see the unexpected behavior.

**Comment by mhegazy** [comment]
I think you are asking about a "solution" file. which is something that does not exist. 

A tsconfig.json is a project, a project is loaded when you open one of its files. if you open module1, module2, biz, foo and bar, you will get all symbols in all projects.

The assumption that opening a folder will open a project, i.e. a single tsconfig.json. 

you can find related suggestions in https://github.com/Microsoft/TypeScript/issues/3469

**Comment by ashelley** [comment]
I'm not really asking for a solution file.  What should happen is:

When you select a file that the type checker and the symbol lookup use the same logic.  

ie: when you select a file where the symbols are loaded from should change (like how the type checker works) to avoid confusion (false positives and missing symbols)

The main problem is that partially supporting multiple tsconfig files in the work place produces very confusing results in regard to symbol lookup.
